front/lib/auth.ts

```diff
@@ -358,6 +364,62 @@ export class Authenticator {
     });
   }
 
+  /**
+   * Exchanges an Authenticator associated with a system key for one associated with a user.
+   *
+   * /!\ This function should only be used with Authenticators that are associated with a system key.
+   *
+   * @param auth
+   * @param param1
+   * @returns
+   */
+  async exchangeSystemKeyForUserAuthByEmail(
+    auth: Authenticator,
+    { userEmail }: { userEmail: string }
+  ): Promise<Authenticator | null> {
+    if (!auth.isSystemKey()) {
+      throw new Error("Provided authenticator does not have a system key.");
+    }
+
+    const owner = auth.workspace();
+    if (!owner) {
+      throw new Error("Workspace not found.");
+    }
+
+    const user = await User.findOne({
+      where: {
+        email: userEmail,
+      },
+    });
+    // If the user does not exist (e.g., whitelisted email addresses),
+    // simply ignore and return null.
+    if (!user) {
+      return null;
+    }
+
+    // Verify that the user has an active membership in the specified workspace.
+    const activeMembership =
+      await MembershipResource.getActiveMembershipOfUserInWorkspace({
+        user: renderUserType(user),
+        workspace: owner,
+      });
+    // If the user does not have an active membership in the workspace,
+    // throw an error to alert the issue.
+    if (!activeMembership) {
+      throw new Error("User not found.");
```

Comment:
This will 500 but really does not look like it :) `null` ?

Reply:
I understand it would depart from your comment. If you really want to fail let's use Result + apiError :pray: